File: python3/step1_read_print.py
import readline  # so input() uses editable input
import reader
import printer


def READ(line):
    return reader.read_str(line)


def EVAL(form):
    return form


def PRINT(line):
    return printer.pr_str(line)


def rep(line):
    return PRINT(EVAL(READ(line)))


def mal():
    print("MAL in Python3")
    print()

    while True:
        try:
            line = input("user> ")
        except EOFError:
            print("\n\nBye")
            break
        result = rep(line)
        print(result)

if __name__ == '__main__':
    mal()

File: java/src/main/java/mal/step1_read_print.java
package mal;

import java.io.Console;

import mal.types.MalException;
import mal.types.MalType;

public class step1_read_print {
  public static void main(String args[]) {
    Console console = System.console();
    String input;
    String output;
    
    while (true) {
      input = console.readLine("user> ");
      if (input == null) {      // Test for EOF
        break; 
      }
      else {
        try {
          output = rep(input);
        }
        catch(MalException ex) {
          output = "*** Error *** " + ex.getMessage();
        }
      }
      System.out.println(output);
    }
  }

  public static MalType READ(String arg) throws MalException {
    return reader.read_str(arg);
  }

  public static MalType EVAL(MalType arg) {
    return arg;
  }

  public static String PRINT(MalType arg) {
    return printer.pr_str(arg);
  }

  public static String rep(String arg) throws MalException {
    String result;

    result = PRINT(EVAL(READ(arg)));
    return result;
  }
}
